src/Features/LanguageServer/Protocol/Handler/OnAutoInsert/OnAutoInsertHandler.cs

```diff
@@ -58,20 +95,83 @@ public async Task<LSP.DocumentOnAutoInsertResponseItem[]> HandleRequestAsync(LSP
 
             if (result == null)
             {
-                return response.ToArray();
+                return null;
             }
 
-            response.Add(new LSP.DocumentOnAutoInsertResponseItem
+            return new LSP.DocumentOnAutoInsertResponseItem
             {
                 TextEditFormat = LSP.InsertTextFormat.Snippet,
                 TextEdit = new LSP.TextEdit
                 {
                     NewText = result.SnippetText.Insert(result.CaretOffset, "$0"),
                     Range = ProtocolConversions.TextSpanToRange(result.SpanToReplace, sourceText)
                 }
-            });
+            };
+        }
+
+        private async Task<LSP.DocumentOnAutoInsertResponseItem?> GetBraceCompletionAfterReturnResponseAsync(LSP.DocumentOnAutoInsertParams autoInsertParams, Document document, CancellationToken cancellationToken)
+        {
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var linePosition = ProtocolConversions.PositionToLinePosition(autoInsertParams.Position);
+            var position = sourceText.Lines.GetPosition(linePosition);
+
+            var serviceAndContext = await GetBraceCompletionContextAsync(position, document, cancellationToken).ConfigureAwait(false);
+            if (serviceAndContext == null)
+            {
+                return null;
+            }
+
+            var (service, context) = serviceAndContext.Value;
+            var postReturnEdit = await service.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: false, cancellationToken).ConfigureAwait(false);
```

Comment:
here is where we'd get the result back, then determine the whitespace to insert due to virtual space.